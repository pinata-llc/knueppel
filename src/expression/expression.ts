import { Query, QueryBuilder } from "../queryBuilder";

export abstract class Expression {
  public abstract build(qb: QueryBuilder): Promise<Query>;
}
