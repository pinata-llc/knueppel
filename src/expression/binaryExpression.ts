import Knex from "knex";
import { ASTNode, ASTParam } from "../ast";
import { UnknownBinaryOperator } from "../error/unknownBinaryOperator";
import { IdentifierResolver } from "../node";
import { Expression } from "./expression";

const binaryOperators = ["=", "!=", "<", "<=", ">", ">=", "+", "-", "*", "/", "%", "in"] as const;
export type BinaryOperator = typeof binaryOperators[number];

@ASTNode
export class BinaryExpression extends Expression {
  constructor(
    @ASTParam("left") protected left: Expression,
    @ASTParam("operator") protected operator: BinaryOperator,
    @ASTParam("right") protected right: Expression,
  ) {
    super();
  }

  public compile(knex: Knex, resolve: IdentifierResolver) {
    const left = this.left.compile(knex, resolve);
    const right = this.right.compile(knex, resolve);

    if (binaryOperators.includes(this.operator)) {
      return knex.raw(`(? ${this.operator} ?)`, [left, right]);
    }

    throw new UnknownBinaryOperator(this.operator);
  }
}
