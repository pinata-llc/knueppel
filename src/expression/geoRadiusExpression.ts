import { ASTNode, ASTParam } from "zargon";
import { QueryBuilder } from "../queryBuilder";
import { Expression } from "./expression";

export type DistanceUnit = "km" | "mi";

const factors = {
  km: 1000,
  mi: 1609.34,
};

@ASTNode
export class GeoRadiusExpression extends Expression {
  constructor(
    @ASTParam("pointA")
    protected pointA: Expression,
    @ASTParam("radius")
    protected radius: Expression,
    @ASTParam("unit")
    protected unit: DistanceUnit,
    @ASTParam("pointB")
    protected pointB: Expression,
  ) {
    super();
  }

  public async build(qb: QueryBuilder) {
    return qb.query(`earth_distance(?, ?) <= ?::numeric * ?::numeric`, [
      await this.pointA.build(qb),
      await this.pointB.build(qb),
      await this.radius.build(qb),
      factors[this.unit],
    ]);
  }
}
