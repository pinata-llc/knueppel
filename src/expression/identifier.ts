import { ASTNode, ASTParam } from "zargon";
import { QueryBuilder } from "../queryBuilder";
import { Expression } from "./expression";

@ASTNode
export class Identifier extends Expression {
  constructor(@ASTParam("name") public name: string, @ASTParam("args") protected args?: Expression[]) {
    super();
  }

  public async build(qb: QueryBuilder) {
    try {
      const identifier = await qb.resolve(this.name, this.args);
      if (!identifier) {
        throw new Error(`Kn√ºppel: Unknown identifier ${identifier}`);
      }
      return qb.query(identifier.queryString, identifier.bindings || [], identifier.tables);
    } catch (err) {
      /* tslint:disable-next-line */
      console.log(err);
      throw err;
    }
  }
}
