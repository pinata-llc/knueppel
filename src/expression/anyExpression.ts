import { ASTNode, ASTParam } from "../index";
import { QueryBuilder } from "../queryBuilder";
import { Expression } from "./expression";
import { Identifier } from "./identifier";
import { ListLiteral } from "./listLiteral";

@ASTNode
export class AnyExpression extends Expression {
  constructor(@ASTParam("values") protected values: ListLiteral | Identifier) {
    super();
  }

  public async build(qb: QueryBuilder) {
    if (this.values instanceof ListLiteral) {
      const query = await this.values.compileAsArray(qb);
      return qb.query("any(?)", [query]);
    }
    return qb.query("any(?)", [await this.values.build(qb)]);
  }
}
