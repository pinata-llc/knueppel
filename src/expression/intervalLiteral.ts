import { ASTNode, ASTParam } from "../ast";
import { UnknownIntervalUnit } from "../error/unknownIntervalUnit";
import { QueryBuilder } from "../queryBuilder";
import { Expression } from "./expression";

const intervalUnits = ["years", "months", "weeks", "days", "hours", "minutes", "seconds"] as const;
type IntervalUnit = typeof intervalUnits[number];

@ASTNode
export class IntervalLiteral extends Expression {
  constructor(
    @ASTParam("value")
    protected value: number,
    @ASTParam("unit")
    protected unit: IntervalUnit,
  ) {
    super();
  }

  public async build(qb: QueryBuilder) {
    if (!intervalUnits.includes(this.unit)) {
      throw new UnknownIntervalUnit(this.unit);
    }

    return qb.query("INTERVAL ?", [`${this.value} ${this.unit}`]);
  }
}
