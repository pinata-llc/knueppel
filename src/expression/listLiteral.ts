import { ASTNode, ASTParam } from "zargon";
import { QueryBuilder } from "../queryBuilder";
import { Expression } from "./expression";

@ASTNode
export class ListLiteral extends Expression {
  constructor(@ASTParam("values", true) protected values: string[] | number[] | boolean[]) {
    super();
  }

  public async build(qb: QueryBuilder) {
    if (Array.isArray(this.values) && this.values.length > 0) {
      const type = typeof this.values[0];
      let arrayType = "";

      if (type === "boolean") {
        arrayType = "::boolean[]";
      } else if (type === "string") {
        arrayType = "::text[]";
      } else if (type === "number") {
        // TODO: handle decimals
        arrayType = "::int[]";
      }
      const marks: unknown[] = this.values;
      return qb.query(`array[${marks.map(() => "?").join(",")}]${arrayType}`, this.values);
    }
    return qb.query("'{}'", []);
  }

  public async buildAsList(qb: QueryBuilder) {
    const marks: unknown[] = this.values;
    if (!Array.isArray(marks) || marks.length === 0) {
      return qb.query("(null)", []);
    }
    return qb.query(`(${marks.map(() => "?").join(",")})`, this.values);
  }
}
