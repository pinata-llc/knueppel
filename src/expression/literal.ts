import { ASTNode, ASTParam } from "../ast";
import { QueryBuilder } from "../queryBuilder";
import { Expression } from "./expression";

@ASTNode
export class Literal extends Expression {
  constructor(@ASTParam("value") protected value: string | number | boolean) {
    super();
  }

  public async build(qb: QueryBuilder) {
    return qb.query("?", [this.value]);
  }
}
