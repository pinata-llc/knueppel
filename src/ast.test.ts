import { Expression } from "./expression/expression";
import { build } from "./index";
import { queryBuilder } from "./utils/test_helpers";

// tslint:disable-next-line
const testJson = require("./ast.test.json");

it("should build query object from a JSON AST", async () => {
  const ast: Expression = build(testJson);
  const query = await ast.build(queryBuilder);
  expect(query.compile()).toMatchSnapshot();
});
