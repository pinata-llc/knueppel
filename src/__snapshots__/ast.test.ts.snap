// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build query object from a JSON AST 1`] = `
Object {
  "bindings": Array [
    "a",
    "a",
    "b",
    "c",
    false,
  ],
  "queryString": "((? = any(array[?,?,?]::text[])) AND not (?))",
}
`;
